python/paddle/primitive/lowering.py

```diff
@@ -0,0 +1,218 @@
+    prepare_python_api_arguments,
+)
+
+_prim_logger = get_logger(
+    "prim", logging.INFO, fmt='%(asctime)s-%(levelname)s: %(message)s'
+)
+
+
+def decomp(
+    block,
+    blacklist=frozenset(),
+    whitelist=frozenset(),
+):
+    """Search nonbasic ops which have be registered composite rules and replace them with primitive ops.
+    The operators in blacklist will be excluded from program when lowering into primitives, and only the
+    operators in whitelist will be lowering. The priority of blacklist is higher than whitelist, it means
+    an operator both in blacklist and whitelist will not be lowering.
+
+    The finally set that will be lowering is:
+        (block.ops & ops have decomposite rule & whitelist) - blacklist
+
+    Args:
+        block (Block|list(Block)): If list, all block must belong to the same program.
+        blacklist (frozenset): The Operators that will be exclude when lowering into primitives.
+        whitelist (frozenset): Only the operators in whitelist will be lowering into primitives.
+    """
+    if isinstance(block, paddle.fluid.framework.Block):
+        main_program = block.program
+    elif isinstance(block, typing.Sequence):
+        for item in block:
+            if not isinstance(item, paddle.fluid.framework.Block):
+                raise TypeError(
+                    f"Expect block or sequence of block, but sequence contains {type(item)}."
+                )
+        main_program = block[0].program
+    else:
+        raise TypeError(
+            f"Expect block or sequence of block, but got {type(block)}."
+        )
+    if not isinstance(blacklist, (set, frozenset)):
+        raise TypeError(
+            f'Expected type of blacklisst is set|frozenset, but got {type(blacklist)}.'
+        )
+    if not isinstance(whitelist, (set, frozenset)):
+        raise TypeError(
+            f'Expected type of whiltelist is set|frozenset, but got {type(whitelist)}.'
+        )
+
+    blacklist = prim_config["forward_blacklist"] | blacklist
+
+    with framework.program_guard(main_program):
+        _prim_logger.info("Decompose composite forward ops begin...")
+
+        if len(blacklist) > 0 and len(whitelist) > 0:
+            op_filter = (
+                lambda x: x.type in whitelist and x.type not in blacklist
+            )
+        elif len(blacklist) > 0 and len(whitelist) == 0:
+            op_filter = lambda x: x.type not in blacklist
+        elif len(blacklist) == 0 and len(whitelist) > 0:
+            op_filter = lambda x: x.type in whitelist
+        else:
+            op_filter = lambda x: True
+        _decom_execute(
+            block,
+            op_filter,
+        )
+        replace_ops = prim_config["composite_ops_record"]
+        _prim_logger.info(
+            f"Decompose composite forward ops finish: {replace_ops}"
+        )
+
+
+def _decom_execute(block, op_filter):
+    """The operators in block wich satisfy the filter conditon will be decomposed into primitives."""
```

Comment:
`decom_execute` is still unreadable，consider how to spilit into different function.

Reply:
done